File: server/internal/usecase/web/dto/instance.go
package dto

import "time"

type InstanceListQuery struct {
	Page    int    `form:"page" validate:"omitempty,min=1"`
	PerPage int    `form:"per_page" validate:"omitempty,min=1,max=100"`
	Status  string `form:"status" validate:"omitempty,oneof=creating running stopped error releasing released"`
}

type InstanceItem struct {
	InstanceNo              string               `json:"instance_no"`
	OrderNo                 string               `json:"order_no"`
	Status                  string               `json:"status"`
	ProductName             string               `json:"product_name"`
	PlanName                string               `json:"plan_name"`
	RegionName              string               `json:"region_name"`
	NetworkTypeName         *string              `json:"network_type_name"`
	TemplateName            string               `json:"template_name"`
	ServiceStartedAt        *time.Time           `json:"service_started_at"`
	ExpiresAt               *time.Time           `json:"expires_at"`
	ExpireStatus            string               `json:"expire_status"`
	ReleaseCountdownSeconds *int64               `json:"release_countdown_seconds"`
	LatestRenewalOrder      *RenewalOrderSummary `json:"latest_renewal_order"`
	CreatedAt               time.Time            `json:"created_at"`
	ReleasedAt              *time.Time           `json:"released_at"`
}

type InstanceDetail struct {
	InstanceItem
	ProductNo                string              `json:"product_no"`
	PlanNo                   string              `json:"plan_no"`
	CPUCores                 int                 `json:"cpu_cores"`
	MemoryMB                 int                 `json:"memory_mb"`
	SystemDiskGB             int                 `json:"system_disk_gb"`
	DataDiskGB               int                 `json:"data_disk_gb"`
	BandwidthMbps            int                 `json:"bandwidth_mbps"`
	RegionNo                 string              `json:"region_no"`
	NetworkTypeNo            *string             `json:"network_type_no"`
	TemplateNo               string              `json:"template_no"`
	OSFamily                 string              `json:"os_family"`
	OSDistribution           string              `json:"os_distribution"`
	OSVersion                string              `json:"os_version"`
	ExpireNoticeSentAt       *time.Time          `json:"expire_notice_sent_at"`
	ExpireReleaseScheduledAt *time.Time          `json:"expire_release_scheduled_at"`
	ExpireReleasedAt         *time.Time          `json:"expire_released_at"`
	RenewalAvailable         bool                `json:"renewal_available"`
	Operations               []InstanceOperation `json:"operations"`
}

type RenewalOrderSummary struct {
	OrderNo          string     `json:"order_no"`
	Status           string     `json:"status"`
	PaymentStatus    string     `json:"payment_status"`
	BillingCycle     string     `json:"billing_cycle"`
	TotalAmountCents uint64     `json:"total_amount_cents"`
	Currency         string     `json:"currency"`
	PaidAt           *time.Time `json:"paid_at"`
	CreatedAt        time.Time  `json:"created_at"`
}

type InstanceOperation struct {
	OperationNo string     `json:"operation_no"`
	Action      string     `json:"action"`
	Status      string     `json:"status"`
	CreatedAt   time.Time  `json:"created_at"`
	CompletedAt *time.Time `json:"completed_at"`
}

File: web/src/api/instance.ts
import { request, type WebApiEnvelope } from './request'
import type { OrderDetail, PageResponse } from './order'

export type InstanceStatus = 'creating' | 'running' | 'stopped' | 'error' | 'releasing' | 'released'
export type InstanceOperationAction = 'provision' | 'start' | 'stop' | 'release' | 'sync'
export type InstanceOperationStatus = 'running' | 'succeeded' | 'failed'

export interface InstanceItem {
  instance_no: string
  order_no: string
  status: InstanceStatus
  product_name: string
  plan_name: string
  region_name: string
  network_type_name: string | null
  template_name: string
  service_started_at: string | null
  expires_at: string | null
  expire_status: 'active' | 'expired' | 'released' | 'unknown'
  release_countdown_seconds: number | null
  latest_renewal_order: RenewalOrderSummary | null
  created_at: string
  released_at: string | null
}

export interface InstanceOperation {
  operation_no: string
  action: InstanceOperationAction
  status: InstanceOperationStatus
  created_at: string
  completed_at: string | null
}

export interface InstanceDetail extends InstanceItem {
  product_no: string
  plan_no: string
  cpu_cores: number
  memory_mb: number
  system_disk_gb: number
  data_disk_gb: number
  bandwidth_mbps: number
  region_no: string
  network_type_no: string | null
  template_no: string
  os_family: string
  os_distribution: string
  os_version: string
  expire_notice_sent_at: string | null
  expire_release_scheduled_at: string | null
  expire_released_at: string | null
  renewal_available: boolean
  operations: InstanceOperation[]
}

export interface RenewalOrderSummary {
  order_no: string
  status: string
  payment_status: string
  billing_cycle: string
  total_amount_cents: number
  currency: string
  paid_at: string | null
  created_at: string
}

export async function getInstances(params?: Record<string, unknown>) {
  const response = await request.get<WebApiEnvelope<PageResponse<InstanceItem>>>('/instances', { params })
  return response.data.data
}

export async function getInstanceDetail(instanceNo: string) {
  const response = await request.get<WebApiEnvelope<InstanceDetail>>(`/instances/${instanceNo}`)
  return response.data.data
}

export async function startInstance(instanceNo: string) {
  const response = await request.post<WebApiEnvelope<InstanceDetail>>(`/instances/${instanceNo}/start`)
  return response.data.data
}

export async function stopInstance(instanceNo: string) {
  const response = await request.post<WebApiEnvelope<InstanceDetail>>(`/instances/${instanceNo}/stop`)
  return response.data.data
}

export async function createRenewalOrder(instanceNo: string, billingCycle: string, clientToken: string) {
  const response = await request.post<WebApiEnvelope<OrderDetail>>(`/instances/${instanceNo}/renewal-orders`, {
    billing_cycle: billingCycle,
    client_token: clientToken,
  })
  return response.data.data
}
